compiler/src/dotty/tools/dotc/report.scala

```diff
@@ -67,11 +67,14 @@ object report:
     error(ex.toMessage, pos, sticky = true)
     if ctx.settings.YdebugTypeError.value then ex.printStackTrace()
 
-  def errorOrMigrationWarning(msg: Message, pos: SrcPos = NoSourcePosition,
-      from: SourceVersion)(using Context): Unit =
-    if sourceVersion.isAtLeast(from) then
-      if sourceVersion.isMigrating && sourceVersion.ordinal <= from.ordinal then migrationWarning(msg, pos)
-      else error(msg, pos)
+  def errorOrMigrationWarning(msg: Message, pos: SrcPos = NoSourcePosition, from: SourceVersion)(using Context): Unit =
+    gradualErrorOrMigrationWarning(msg, pos, from, from)
+
+  def gradualErrorOrMigrationWarning(msg: Message, pos: SrcPos = NoSourcePosition, warnFrom: SourceVersion, errorFrom: SourceVersion)(using Context): Unit =
+    if sourceVersion.isAtLeast(warnFrom) then
+      if sourceVersion.isMigrating && sourceVersion.ordinal <= errorFrom.ordinal then migrationWarning(msg, pos)
```

Comment:
I think this only makes sense if the source version is migrating to the errorFrom version:
```suggestion
      if sourceVersion.isAtLeast(errorFrom) then
        if sourceVersion.isMigrating && sourceVersion.ordinal <= errorFrom.ordinal then migrationWarning(msg, pos)
        else error(msg, pos)
      else warning(msg, pos)
```

Reply:
so this probably makes more sense to keep `errorOrMigrationWarning` the same, and `gradualErrorOrMigrationWarning` wraps it

Reply:
I guess I don't have a clear understanding of the criteria for when a migration warning should be emitted vs. a regular warning.

My line of thinking was that if the current source version is `$version-migration`, and the diagnostic concerns language features or changes introduced by $version (and perhaps offers rewrites, etc.), then it should be a migration warning.

My first inclination in this case was to emit a migration warning under both `3.2` and `3.2-migration`, but it seemed that the current usage of migration warnings is limited to migration source versions.

I don't have strong feelings on this either way, but I think it might be good to document the consensus opinion to avoid any future misinterpretation.

Reply:
from what I understand from the status-quo, migration warning is to turn errors in `3.0` into migration warnings if you use `3.0-migration`. So then once you go beyond the stable version, it is always an error. (i.e. it doesn't become again a migration warning in e.g. `3.3-migration`)

Reply:
done